src/bucketizers/timeBucketTree.ts

```diff
@@ -0,0 +1,377 @@
+            value: x.id.value,
+            literal: x.id,
+        }));
+
+        if (save) {
+            this.state = JSON.parse(save);
+            hydrate(this.state);
+        }
+    }
+
+    bucketize(
+        sdsMember: Record,
+        getBucket: (key: string, root?: boolean) => Bucket,
+        addRelation: AddRelation,
+    ): Bucket[] {
+        const values = this.path
+            .execute(sdsMember.data)
+            .filter(
+                (x, i, arr) => arr.findIndex((y) => x.value === y.value) == i,
+            );
+
+        const out: Bucket[] = [];
+        for (const value of values) {
+            const date = new Date(value.value);
+            const endDate = new Date(date.getTime() - this.config.timeBufferMs);
+            let key = "";
+            let state = this.state;
+            let bucket = getBucket(key, true);
+            for (const level of this.config.levels) {
+                checkImmutable(state, key, endDate, getBucket);
+
+                const levelValue = level.ranges
+                    .map((x) => levelToValue[x](date))
+                    .join("-");
+
+                const lastF = (date: Date) =>
+                    level.ranges.reduce(
+                        (date, level) => levelMax[level](date),
+                        date,
+                    );
+
+                const minF = (date: Date) =>
+                    level.ranges.reduce(
+                        (date, level) => levelMin[level](date),
+                        date,
+                    );
+
+                const found = goInState(state, levelValue, date, lastF);
+
+                state = found.value.deep;
+                key = concat_key(key, levelValue);
+
+                const nextBucket = getBucket(key);
+                if (!found.found) {
+                    const minDate = minF(date).toISOString();
+                    const maxDate = lastF(date).toISOString();
+
+                    addRelation(
+                        bucket,
+                        nextBucket,
+                        TREE.terms.GreaterThanOrEqualToRelation,
+                        literal(minDate, namedNode(XSD.dateTime)),
+                        this.config.pathQuads,
+                    );
+
+                    addRelation(
+                        bucket,
+                        nextBucket,
+                        TREE.terms.LessThanRelation,
+                        literal(maxDate, namedNode(XSD.dateTime)),
+                        this.config.pathQuads,
+                    );
+                }
+                bucket = nextBucket;
+                if (found.value.count < level.amount) {
+                    found.value.count += 1;
+                    break;
+                }
+            }
+
+            out.push(bucket);
+        }
+
+        return out;
+    }
+
+    save(): string {
+        return JSON.stringify(this.state);
+    }
+}
+
+function concat_key(path: string, key: string): string {
```

Comment:
Can I vote for using camelCase everywhere?

Reply:
You can vote, but then we need a vote :eyes: